### PR TITLE
🌐 (frontend) Add internationalization

### PR DESCRIPTION
**Pull request by AntoLC**
## Purpose

Add internationalization.

## Proposal

- [x] install and setup packages to Desk (i18next)
- [x] make all the string fields translatable
- [x] Plug the Language Picker with i18next
- [x] Plug it with crowdin
- [x] Test  and checks with CI if frontend translations are made

## How to use on the frontside

```ts
import { useTranslation } from 'react-i18next';

export default function Home() {
  const { t } = useTranslation();

  return (
    <div>
      <h1>{t('Hello Desk !')}</h1>
      <h2>{t('Hello {{count}} !', { count: 2 })}</h2> {/* Will display Hello 2 ! */}
      <h2>
         {t('Hello world !', {
          description: 'Fields to describe about what it is in crowdin.',
         })}
      </h2>
    </div>
  );
}
```

The keys are in english, we can skip the english translation like that, but keep in mind that there are still keys, so we can add a english translation above them.

### Crowdin Init
Steps:
- Get your access to the project on https://crowdin.com, copy the project id
- Set the var in `env.d/development/crowdin` (`CROWDIN_API_TOKEN `/ `CROWDIN_PROJECT_ID`)

### Send translation to crowdin
Back and front: `make i18n-generate-and-upload`
Only frontend: `make frontend-i18n-generate && make crowdin-upload`

### Pull translation from crowdin
Back and front: `make i18n-download-and-compile`
Only frontend: `make crowdin-download && make frontend-i18n-compile`

⚠️ If you get a warning during your pull, it is probably because you need to have init at least 1 times crowdin CLI before to be able to pull the translations, to do so, do `make i18n-generate-and-upload`, then redo your pull.

## Good to know
✅ The CI send automatically translations to Crowdin.
✅  The CI will detect if some translations are missing, it will tell you from the job `test-front/Test Translations`
If this job fails, go to https://crowdin.com, make your translations, then pull your translations and repush your PR.

### CODE REVIEW

File: src/frontend/packages/i18n/package.json [hunk 1]
```diff
@@ -6,10 +6,16 @@
...
+    "format-deploy:desk": "node ./format-deploy.mjs --app=desk --output=../../apps/desk/src/i18n/translations.json",
+    "test": "jest"
   },
   "dependencies": {
+    "@types/jest": "29.5.11",
+    "@types/node": "20.11.6",
     "i18next-parser": "8.8.0",
+    "jest": "29.7.0",
+    "ts-jest": "29.1.2",
+    "typescript": "*",
```
Comment (lebaudantoine) [inline src/frontend/packages/i18n/package.json:21]: I have a question, should we pin this `typescript`? Same question for other package?

File: src/frontend/apps/desk/src/i18n/utils.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  BASE_LANGUAGE,
+  LANGUAGES_ALLOWED,
+  LANGUAGE_LOCAL_STORAGE,
+} from './conf';
+
+export const splitLocaleCode = (language: string) => {
+  const locale = language.split(/[-_]/);
+  return {
+    language: locale[0],
+    region: locale[1],
```
Comment (lebaudantoine) [inline src/frontend/apps/desk/src/i18n/utils.ts:11]: This key can be undefined right ?

File: src/frontend/apps/e2e/__tests__/app-desk/language.spec.ts [hunk 3]
```diff
@@ -18,9 +18,6 @@ test.describe("Language", () => {
     const header = page.locator("header").first();
 
     await header.getByRole("combobox").getByText("FR").click();
-    await expect(
-      header.getByRole("option", { name: "Language Icon FR" }),
-    ).toHaveAttribute("aria-selected", "true");
```
Comment (lebaudantoine) [inline src/frontend/apps/e2e/__tests__/app-desk/language.spec.ts]: question, why did you remove this ?

File: .github/workflows/people.yml [hunk 4]
```diff
@@ -58,24 +58,30 @@ jobs:
...
           node-version: '18.x'
           cache: 'yarn'
           cache-dependency-path: src/frontend/yarn.lock
+
       - name: Install dependencies
         run: yarn install --frozen-lockfile
-      - name: Test Desk App
-        run: yarn test
+
+      - name: Test app
```
Comment (lebaudantoine) [inline .github/workflows/people.yml:80]:
```suggestion
      - name: Test App
```

File: src/frontend/apps/desk/src/i18n/utils.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  const language = splitLocaleCode(languageStore).language;
+
+  if (LANGUAGES_ALLOWED.includes(language)) {
+    return language;
+  }
+
+  return BASE_LANGUAGE;
+};
```
Comment (lebaudantoine) [inline src/frontend/apps/desk/src/i18n/utils.ts:29]:
```suggestion
export const getLanguage = () => {

  if (typeof window == 'undefined') {
    return BASE_LANGUAGE
  }

  const languageStore = localStorage.getItem(LANGUAGE_LOCAL_STORAGE) || navigator?.language;
  const language = splitLocaleCode(languageStore).language;

  if (LANGUAGES_ALLOWED.includes(language)) {
    return language;
  }

  return BASE_LANGUAGE;
};
```

File: src/frontend/packages/i18n/__tests__/i18n.test.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    fs.mkdirSync("./locales/tests/fr/", { recursive: true });
+    fs.writeFileSync(
+      "./locales/tests/fr/translations.json",
+      JSON.stringify({ test: { message: "Mon test" } }),
+      "utf8",
+    );
+    expect(fs.existsSync("./locales/tests/fr/translations.json")).toBeTruthy();
+
+    // execute format-deploy command
```
Comment (lebaudantoine) [inline src/frontend/packages/i18n/__tests__/i18n.test.ts:48]: nit:
```suggestion
    // Execute format-deploy command
```

File: src/frontend/packages/i18n/__tests__/translations.test.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+describe("checks all the frontend translation are made", () => {
+  it("checks missing translation. If this test fails, go to crowdin.com", () => {
+    // Extract the translations
+    execSync("yarn extract-translation:desk");
+    const outputCrowdin = "./locales/desk/translations-crowdin.json";
+    const jsonCrowdin = JSON.parse(fs.readFileSync(outputCrowdin, "utf8"));
+    const listKeysCrowdin = Object.keys(jsonCrowdin).sort();
+
+    // Checks the translations in the app Desk
```
Comment (lebaudantoine) [inline src/frontend/packages/i18n/__tests__/translations.test.ts:12]:
```suggestion
    // Check the translations in the app Desk
```

File: src/frontend/packages/i18n/__tests__/translations.test.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    const outputDesk = "../../apps/desk/src/i18n/translations.json";
+    const jsonDesk = JSON.parse(fs.readFileSync(outputDesk, "utf8"));
+
+    Object.keys(jsonDesk).forEach((key) => {
+      // Our keys are in english, so we don't need to check the english translation
+      if (key === "en") {
+        return;
+      }
```
Comment (lebaudantoine) [inline src/frontend/packages/i18n/__tests__/translations.test.ts:20]:
```suggestion
    // Our keys are in english, so we don't need to check the english translation
    Object.keys(jsonDesk).filter((key) => !== "en").forEach((key) => {
```

**Review by lebaudantoine — COMMENTED** [review]
Code Review ! I need to test the script locally

File: src/frontend/package.json [hunk 9]
```diff
@@ -17,7 +17,11 @@
     "app:test": "yarn APP_DESK run test",
     "ci:build": "yarn APP_DESK run build:ci",
     "e2e:test": "yarn APP_E2E run test",
-    "lint": "yarn APP_DESK run lint && yarn APP_E2E run lint"
+    "lint": "yarn APP_DESK run lint && yarn APP_E2E run lint && yarn workspace eslint-config-people run lint && yarn i18n run lint",
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/frontend/package.json:20]: Let's see that in another PR.

File: src/frontend/packages/i18n/package.json (hunk 1 above)
Comment (AntoLC) [reply src/frontend/packages/i18n/package.json:21]: It is pinned , here: 
https://github.com/numerique-gouv/people/blob/13a58d6fa0b2af52b9b9b4c12053d7eb11cdb528/src/frontend/package.json#L22-L25

To do like that, we have the insurance that all the project is with the same version of certain packages.

File: src/frontend/apps/e2e/__tests__/app-desk/language.spec.ts (hunk 3 above)
Comment (AntoLC) [reply src/frontend/apps/e2e/__tests__/app-desk/language.spec.ts]: It is a good question, it seems to have a bug in the Cunningham select during the first render between `defaultValue` and `options`, the options is not selected, if you click on it another time the option will pass to selected.

File: .gitignore [hunk 10]
```diff
@@ -78,3 +78,4 @@ db.sqlite3
 .idea/
 .vscode/
 *.iml
+.devcontainer/
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply .gitignore]: https://github.com/numerique-gouv/people/pull/17/commits/900a28bff99c6a3028d87c5d36dfd706767ec7b4#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R56

File: src/frontend/apps/desk/.env.test [hunk 11]
```diff
@@ -1 +1 @@
-NEXT_PUBLIC_API_URL=/api/
+NEXT_PUBLIC_API_URL=/api/
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/frontend/apps/desk/.env.test]: https://github.com/numerique-gouv/people/pull/17/commits/df3b394f36cc341a13b28f8d12520902dc7df5f2

File: Makefile [hunk 12]
```diff
@@ -53,7 +53,8 @@ MAIL_YARN           = $(COMPOSE_RUN) -w /app/src/mail node yarn
 TSCLIENT_YARN       = $(COMPOSE_RUN) -w /app/src/tsclient node yarn
 
 # -- Frontend
-PATH_FRONT_DESK       	= ./src/frontend/apps/desk
+PATH_FRONT       		= ./src/frontend
+PATH_FRONT_DESK     = $(PATH_FRONT)/apps/desk
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply Makefile]: https://github.com/numerique-gouv/people/pull/17/commits/900a28bff99c6a3028d87c5d36dfd706767ec7b4#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R56

File: src/frontend/packages/i18n/.eslintrc.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    "plugin:import/recommended",
+  ],
+  parserOptions: {
+    sourceType: "module",
+    ecmaVersion: "latest",
+    tsconfigRootDir: __dirname,
+    project: ['./tsconfig.json'],
+  },
+  ignorePatterns: ['.eslintrc.js'],
+};
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/frontend/packages/i18n/.eslintrc.js]: https://github.com/numerique-gouv/people/pull/17/commits/94a68fb68301f1ba674c6c919c8e1533831a8d9d

File: src/frontend/package.json [hunk 14]
```diff
@@ -17,7 +17,11 @@
...
+    "i18n": "yarn workspace packages-i18n",
```
[the remark this answers is not in the document]
Comment (AntoLC) [reply src/frontend/package.json]: https://github.com/numerique-gouv/people/pull/17/commits/900a28bff99c6a3028d87c5d36dfd706767ec7b4#diff-12ee508ecc1901fe1a1d71ed459dba8454f3160983ee18b2c51b76ade45235d3R14

File: src/frontend/apps/desk/src/i18n/utils.ts (hunk 2 above)
Comment (AntoLC) [reply src/frontend/apps/desk/src/i18n/utils.ts]: https://github.com/numerique-gouv/people/pull/17/commits/df3b394f36cc341a13b28f8d12520902dc7df5f2#diff-4199f89f406d85e6fd929a6d2de6737f8b041b22bbc9903e1f0764287354c8a5R11

File: .github/workflows/people.yml (hunk 4 above)
Comment (AntoLC) [reply .github/workflows/people.yml]: https://github.com/numerique-gouv/people/pull/17/commits/95199a3820cd6c462a8d51ad74b49e0bf77b6a5a

File: src/frontend/apps/desk/src/i18n/utils.ts (hunk 5 above)
Comment (AntoLC) [reply src/frontend/apps/desk/src/i18n/utils.ts]: https://github.com/numerique-gouv/people/pull/17/commits/df3b394f36cc341a13b28f8d12520902dc7df5f2

File: src/frontend/packages/i18n/__tests__/i18n.test.ts (hunk 6 above)
Comment (AntoLC) [reply src/frontend/packages/i18n/__tests__/i18n.test.ts]: https://github.com/numerique-gouv/people/pull/17/commits/95199a3820cd6c462a8d51ad74b49e0bf77b6a5a

File: src/frontend/packages/i18n/__tests__/translations.test.ts (hunk 7 above)
Comment (AntoLC) [reply src/frontend/packages/i18n/__tests__/translations.test.ts]: https://github.com/numerique-gouv/people/pull/17/commits/95199a3820cd6c462a8d51ad74b49e0bf77b6a5a

File: src/frontend/packages/i18n/__tests__/translations.test.ts (hunk 8 above)
Comment (AntoLC) [reply src/frontend/packages/i18n/__tests__/translations.test.ts]: https://github.com/numerique-gouv/people/pull/17/commits/95199a3820cd6c462a8d51ad74b49e0bf77b6a5a

File: src/frontend/apps/e2e/__tests__/app-desk/language.spec.ts (hunk 3 above)
Comment (lebaudantoine) [reply src/frontend/apps/e2e/__tests__/app-desk/language.spec.ts]: Let's open an issue in Cunningham and ignore my comment then.

File: src/frontend/apps/desk/src/app/Language/LanguagePicker.tsx [hunk 15]
```diff
@@ -31,49 +33,42 @@ const SelectStyled = styled(Select)<{ $isSmall?: boolean }>`
...
-        <Image priority src={IconLanguage} alt="Language Icon" />
-        <Text>EN</Text>
-      </Box>
-    ),
-  },
-];
-
 export const LanguagePicker = () => {
+  const { t, i18n } = useTranslation();
+  const { preload: languages } = i18n.options;
```
Comment (lebaudantoine) [inline src/frontend/apps/desk/src/app/Language/LanguagePicker.tsx:38]: I faced some issue while loading translation, and all languages were still accessible in the dropdown. By default, if translation are not loaded properly, we would prefer have a check, preventing the user selecting a language which has not been loaded.

**Comment by lebaudantoine** [comment]
It would be nice to add pulling translations at least once into make bootstrap to eliminate any extra steps for developers

File: src/frontend/apps/desk/src/app/Language/LanguagePicker.tsx (hunk 15 above)
Comment (AntoLC) [reply src/frontend/apps/desk/src/app/Language/LanguagePicker.tsx:38]: We have test to prevent that, you cannot merge a branch when the translations are not made.
https://github.com/numerique-gouv/people/blob/94a68fb68301f1ba674c6c919c8e1533831a8d9d/src/frontend/packages/i18n/__tests__/translations.test.ts#L4

**Comment by AntoLC** [comment]
> It would be nice to add pulling translations at least once into `make bootstrap` to eliminate any extra steps for developers

You don't have to pull the translations when you switch to a branch, the translations are already made in the branch. 
If you do a pulling you will potentially create new changes in the branch not related to the branch you are working with.

File: src/frontend/apps/desk/src/app/Language/LanguagePicker.tsx (hunk 15 above)
Comment (lebaudantoine) [reply src/frontend/apps/desk/src/app/Language/LanguagePicker.tsx:38]: I am not sure your answer, matches my question. I am concerned about the Language selector showing all languages, even if translations have not been pulled from Crowdin. Let's discuss it orally.

**Review by lebaudantoine — APPROVED** [review]

File: src/frontend/apps/desk/src/i18n/utils.ts (hunk 2 above)
Comment (AntoLC) [reply src/frontend/apps/desk/src/i18n/utils.ts]: Like that (`const [language, locale] = lng.split(/[-_]/);`) locale has the type `string`, when actually locale can be `string | undefined`.
